content/4_Gold/LIS.mdx

```diff
@@ -192,6 +192,72 @@ def find_lis(arr: List[int]) -> int:
 
 </LanguageSection>
 
+### Fast Solution (RMQ Data Structures)
+
+(By Aryansh Shrivastava) 
+
+**Time Complexity**: $\mathcal O(N \log N)$, but perhaps a constant factor slower than binary search. 
+
+Note that the following solution assumes knowledge of a basic PURQ data structure for RMQ. Suitable implementations include a segment tree or a modified Fenwick tree, but both of these are generally considered Platinum level topics. However, this solution has the advantage of being more intuitive if you're deriving LIS on the spot. 
+
+Let $\texttt{dp}[k]$ be the length of the LIS of $a$ that ends with the element $k \in a.$  Our base case is obviously $\texttt{dp}[a[0]] = 1$ (we have an LIS containing only $a[0],$ which has length $1$). Since LIS exhibits optimal substructure, we transition as follows, processing $\forall t \in a[1 \dots n-1]$ from left to right (which we must do in order to maintain the property that our subsequence strictly increases in this direction):
```

Comment:
```suggestion
Let $\texttt{dp}[k]$ be the length of the LIS of $a$ that ends with the element $k \in a$. Our base case is obviously $\texttt{dp}[a[0]] = 1$ (we have an LIS containing only $a[0]$, which has length $1$). Since LIS exhibits optimal substructure, we transition as follows, processing $\forall t \in a[1 \dots n-1]$ from left to right (which we must do in order to maintain the property that our subsequence strictly increases in this direction):
```